security/pkg/k8s/tokenreview/k8sauthn.go

```diff
@@ -196,44 +234,48 @@ func (authn *K8sSvcAcctAuthn) ValidateK8sJwt(targetToken, jwtPolicy string) ([]s
 		}
 	}
 	if !inServiceAccountGroup {
-		return nil, fmt.Errorf("the token is not a service account")
+		return nil, fmt.Errorf("found the jwt token in cluster %v, however the token is "+
+			"not associated with a service account", clusterID)
 	}
 	// "username" is in the form of system:serviceaccount:{namespace}:{service account name}",
 	// e.g., "username":"system:serviceaccount:default:example-pod-sa"
 	subStrings := strings.Split(tokenReview.Status.User.Username, ":")
 	if len(subStrings) != 4 {
-		return nil, fmt.Errorf("invalid username field in the token review result")
+		return nil, fmt.Errorf("found the jwt token in cluster %v, however found invalid username field "+
+			"in the token review result", clusterID)
 	}
 	namespace := subStrings[2]
 	saName := subStrings[3]
-
 	return []string{namespace, saName}, nil
 }
 
-// isTrustworthyJwt checks if a jwt is a trustworthy jwt type.
-func isTrustworthyJwt(jwt string) (bool, error) {
-	type trustWorthyJwtPayload struct {
-		Aud []string `json:"aud"`
-		Exp int      `json:"exp"`
-	}
-
-	jwtSplit := strings.Split(jwt, ".")
-	if len(jwtSplit) != 3 {
-		return false, fmt.Errorf("jwt may be invalid: %s", jwt)
-	}
-	payload := jwtSplit[1]
-
-	payloadBytes, err := base64.RawStdEncoding.DecodeString(payload)
-	if err != nil {
-		return false, fmt.Errorf("failed to decode jwt: %v", err.Error())
+//ValidateRemoteK8sJwt validates against a remote k8s JWT at API server.
+func ValidateRemoteK8sJwt(targetJWT, jwtPolicy string, mc *kubecontroller.Multicluster) ([]string, error) {
+	tokenReview := &k8sauth.TokenReview{}
+	if jwtPolicy == jwt.JWTPolicyThirdPartyJWT {
+		tokenReview.Spec = k8sauth.TokenReviewSpec{
+			Token:     targetJWT,
+			Audiences: []string{defaultAudience},
+		}
+	} else if jwtPolicy == jwt.JWTPolicyFirstPartyJWT {
+		tokenReview.Spec = k8sauth.TokenReviewSpec{
+			Token: targetJWT,
+		}
+	} else {
+		return nil, fmt.Errorf("invalid JWT policy: %v", jwtPolicy)
 	}
-
-	structuredPayload := &trustWorthyJwtPayload{}
-	err = json.Unmarshal(payloadBytes, &structuredPayload)
-	if err != nil {
-		return false, fmt.Errorf("failed to unmarshal jwt: %v", err.Error())
+	remoteClients := mc.GetRemoteKubeClients()
+	for k, v := range remoteClients {
+		reviewRes, err := v.AuthenticationV1().TokenReviews().Create(context.TODO(), tokenReview, metav1.CreateOptions{})
+		if err != nil {
```

Comment:
I think when error occurs, we should not do retry in this cluster.

Reply:
will `apierrors.IsInternalError(err) || apierrors.IsServerTimeout(err) || apierrors.IsTimeout(err)` cover all the cases?

Reply:
I think we should do retries in these cases. The only case that we should stop retrying is when the cluster is returning a validation error (which means the JWT isn't signed by itself). And the validation error may be specific on 1) the JWT is not signed by me 2) the JWT is signed by me, but the service account is removed. If we can distinguish them based on the error message, we should continue to next remoteClient when (1) happens, and we should break this loop when (2) happens.

When we have deterministic mapping between the JWT and the clusters. We can do better retries here and greatly reduce the overhead due to the errors.

Does this make sense?

Reply:
added retry logic for some errors